### ISSUE TITLE
Integrate WhereYouGo into c:geo

### ISSUE THREAD

**Issue by eddiemuc** [cgeo/cgeo#10087]
Was not sure whether to file this issue in WhereYouGo project or in c:geo project.

Two major things come to my mind which would need upgrading in WhereYouGo in the near future:
* mapsforge
* SAF

Browsing a bit into WhereYouGo, I couldn't help but thinking that it might actually be less work to port it to c:geo (not OpenWig, just WhereYouGo) rather than investing into WhereYouGo upgrade and future maintenance. I would be interested to get some opinions about this.

From technical side: I calculated some very rough statistics. It seems WhereYouGo has about 11K LoCs (c:geo roughly 84K), 2,6K of this alone is related to maps and thus somehow duplicated to c:geo. Of course the majority of Code is GUI (about 4K), but I assume that there are a lot of base classes inside this which I assume c:geo can provide as well.

**Comment by geocachermgo** [comment]
I was not aware that WIG is maintained by the same team. But if you think that even the next unaviodable changes are more effort then integrating it, then on the longer run the advantage for integration should be even higher.

**Comment by Narrat** [comment]
It has its merits if both are separate. At least for my way of using it.
Could do a wherigo and look for other Caches on the way.
If it were to be included it would kinda block the interface? At least it would involve more interaction as compared to a simple switch between two apps.
But I do understand that it is cumbersome to the same task more than once. And maybe such an integration would be a lot smoother to play with (compared to what I currently think of it)

**Comment by MagpieFourtyTwo** [comment]
Would of course be great if WYG could be integrated into c:geo, esp. as it seems to be not as good maintained as c:geo.
Moreover doing a WIG could possibly get a somewhat cleaner user interface this way,
and an easier entry for new users (setting up WYG is still a pain in the a...).
But OTOH I see c:geo, loaded with tons of features - which may affect stability ...?  And make it harder to maintain ...?

> It has its merits if both are separate. At least for my way of using it.
> Could do a wherigo and look for other Caches on the way.

While doing a Multi you can look for other caches, too - and I would expect WIGs to be handled in a comparable way:
While going to the next WIG area, you are in c:geo's map, where you should be able to activate "Live".
And if so, you would not even have to switch context to look for other caches ...

**Comment by Lomanic** [comment]
Previous feature request to have a whereigo player in cgeo in #7717 was rejected as too much work.

**Comment by Lineflyer** [comment]
@Lomanic Back at those days we did not have the freedom to reuse WhereYouGo source code, but meanwhile it 1. has the appropriate license and 2. is in our hands.

In general about the idea:
I do very well understand the reasoning to integrate WhereYouGo and it would be really nice. However the major challenge IMHO is how a UI would look like and how it would integrate and interwork with the remaining app.
I do also have the concern, similar to @Narrat , that such implementation should not block the remaining app while playing. This needs a really clever implementation, we should sketch first before finally deciding upon integration.

**Comment by eddiemuc** [comment]
@Lomanic actually, since WhereYouGo somehow seems to be also in responsibility of the c:geo team, we have to ask ourself what is more work:
* integrate WYG into c:geo and continue maintaining it there
* Leaving it as a separate app, make mandatory changes (upgrade mapsforge version, make ready for scoped storage) and continue maintaining the app in parallel to c:geo

I did not list the option "drop support for WYG" because I hope this is not seriously considered. In that case I would need to switch to "Geooh Go" which has an integrated WYG player - and I really don't want that

**Comment by eddiemuc** [comment]
In PR #10232 I did a first (and not yet working!) move of WhereYouGo code to c:geo. 
There's a lot of work to be done, but frankly it is less than was necessary for something like SAF (and definitely less than maintaining two times map features in the long run!). 

Could look something like this:

![image](https://user-images.githubusercontent.com/6909759/111912110-79d27680-8a68-11eb-8661-522646e50fe8.png)

Should I pursue this path?

**Comment by UniQP** [comment]
What about moving the common code into a library that is then used by both apps? Integrating WhereYouGo into c:geo feels wrong to me, because the apps have different purposes and the code would be harder to maintain afterwards.

**Comment by eddiemuc** [comment]
> What about moving the common code into a library that is then used by both apps? Integrating WhereYouGo into c:geo feels wrong to me, because the apps have different purposes and the code would be harder to maintain afterwards.

I like the idea of having one app where I can see all Geocache Data on one integrated map (caches, labs, wherigos, trackables, routing etc). I think this fits together. Geeoh is also doing it this way.

Wrt integrated code, this will be less/comparable to e.g. integration of BRouter, so I guess that's ok. 

Maintaining a three-way-structure (lib + 2 apps) in my personal experience creates far more maintenance work than one integrated app.

 However, if it is felt that this is the wrong way, I can stop approaching this path immediately. This is why I ask after all :-)

**Comment by bekuno** [comment]
I vote for a separate WhereYouGo app.
1) c:geo is not the only geocaching app on Android devices. But WhereYouGo is. 
2) I would suggest an implementation like with Locus. There you can see both caches and WhereYouGo regions.
![Screenshot_20210321-184945_Locus Map Pro 4](https://user-images.githubusercontent.com/32555742/111915941-ee151600-8a78-11eb-9d5e-f272bef6bced.jpg)

**Comment by eddiemuc** [comment]
> I vote for a separate WhereYouGo app.
> 1) c:geo is not the only geocaching app on Android devices. But WhereYouGo is. 
> 2) I would suggest an implementation like with Locus. There you can see both caches and WhereYouGo regions.

@bekuno 
I am confused. The second argument (which I share) would imho imply that an integrated app (with both caches and wherigos) is preferrable over two separate apps. What am I understanding wrong here?

(Also I didn't know that Locus supports Wherigos?)

**Comment by bekuno** [comment]
> (Also I didn't know that Locus supports Wherigos?)

Ahh, thats the point.
You can in WhereYouGo Global Settings select "Map provider", either Mapsforge od Locus.
My screenshot was created with the Locus configuration. 
Then the tap of "Map" button stats/opens Locus.
And if you have also the Geocaching4Locus Addon live map active then both are visible.

**Comment by eddiemuc** [comment]
A question came to my mind: it the expressed wish for a "split" of wherigo and other cache types more a technical one (e.g. split into separate apps/libs/repos/developer-teams) or a UX one (e.g. split into separately manageable apps with different maps, settings, UX styles etc)?

Because mixtures would be possible. For example:
* one codebase could be build to support both whereigos and normal caches, but out of this codebase two different apps (with distinct maps etc) could be build (even today, one can put c:geo's main functions into separate starter icons on Android screens)
* Or the other way around: code could be splitted into different libraries (for example map viewer, wherigo, also other parts like SAF, cache list display/filtering etc) but one app could be build out of these different parts

**Comment by eddiemuc** [comment]
> > (Also I didn't know that Locus supports Wherigos?)
> 
> Ahh, thats the point.
> You can in WhereYouGo Global Settings select "Map provider", either Mapsforge od Locus.
> My screenshot was created with the Locus configuration.
> Then the tap of "Map" button stats/opens Locus.
> And if you have also the Geocaching4Locus Addon live map active then both are visible.

If one uses Geocaching4Locus, why would he use c:geo in parallel? 
Or the other way around: if one uses c:geo, why would he in addition use Geocaching4Locus (is there something offered by it which c:geo can't offer?). If the only reason for that is that Whereigos can be displayed in conjunction with normal caches -> lets integrate this possibility in c:geo

**Comment by bekuno** [comment]
Locus is a universal outdoor app. Geocaching is a special part of it and realized as an addon from an other developer than Locus. There are c:geo users which use Locus for navigation to the caches (rotate map, other overlays, maybe more). (I think that they use Geocaching4Locus not as default geocaching app). WhereYouGo was developed from the Locus main developer, so it is understandable that there is the Locus integration.

The main pain in WhereYouGo (beside SAF necessary in future) is the lack of support of mapsforge >v3.
And one indirect solution would be the visibility in c:geo.
This would then solve an other point for c:geo users - view caches and WhereYouGo areas in the same time.

Maybe the direct support for mapsforge v4 in WhereYouGo is then a result of the c:geo integration (dream).

**Comment by moving-bits** [comment]
I have no strong opinion on whether to better integrate a Wherigo player into c:geo or maintain it as a separate app, but I see some pains we have with today's solution, mainly
- having to maintain a separate app (which has some overlapping functionality with c:geo, but implemented in a different way),
- WhereYouGo only supporting the old mapsforge, and
- WhereYouGo not being ready for SAF.

Personally, I could live with both WhereYouGo being a separate app, or a Whereigo player being integrated into c:geo. On the one hand side, having WIG (and maybe ALC, some day) within c:geo would make for a nice "all geocaching in one" app, OTOH we need to make sure to not make the code base too complicated.

So I guess one should decide what's less work to do - integrate it into c:geo, reusing/sharing code from c:geo, or updating WYG to newer Mapsforge, SAF etc. (Maybe the letter could be eased with putting all our SAF stuff into a lib? Don't know if that would be feasible with the map stuff as well - probably not.)

**Comment by eddiemuc** [comment]
Ok let's think about the "lib" approach for a minute from a technical standpoint. 

In my opinion, when approaching this way there would be the following things to share between the apps:
* Preferences framework
* SAF framework (which relies on preferences lib too)
* Map viewer framework (which relies on the other two libs as well)
* Map viewer add-ons (e.g. routing or map download)

Important design decision would be whether those libs shall be "c:geo-internal only" (only used for c:geo and WYG) or should be made available public themselves for other developers to use (like e.g. mapsforge). For the later a much cleaner and we'll documented API is necessary, for the former we might be a bit "lazier" and maybe even put all of the above stuff into one big lib. Opinions on this?

It might be noted that SAF framework in its current form is far too overpowered for most apps, especially for WYG. 

And a wish from my side: in the end I really want to be able to see wherigos stuff and normal caches on one map (not using Locus), so in order to achieve that with a two-app-approach I guess the two apps have to have some sort of sync between each other.

Also note that for a lib-apo-approach we need to redesign a bit our CI such that on changes of a lib the apps are retested, used API level and 3rd party libs stay compatible etc. If we decide to share our new libs with the outside we even need some dedicated test apps to test the libs also outside of c:geo/WYG context (e.g. mapsforge has such a test app and it is absolutely necessary).

In summary, I think this is all doable and a nice piece of software design work, but I really think this is a lot of effort.

**Comment by eddiemuc** [comment]
(added some more things above which are shareable)

**Comment by eddiemuc** [comment]
The question wrt integration of WhereYouGo into c:geo relates undecided. In case of decision for integration into c:geo I volunteer for doing it. In other case I would close this issue and leave WYG be.

**Comment by Lineflyer** [comment]
I think the only chance for WhereYouGo to survive in our repo is to bind it more close together with c:geo. Regarding the argument, that people might want to use WhereYouGo but not c:geo (for whatever reason), I think we can be very pragmatic:
We took the source code to keep it alive and now realized we cannot do that in additon but only if we find synergies. If someone wants to keep it separate he is free to take the source code and maintain and publish it separately.

The key question for me (as I mentioned above) is, how an integration could look like:
The gameplay of an cartridge might require spontaneous or continues interaction, this is something we do not have at this moment in c:geo and I have no idea how it could work UI-wise...

The idea of offering c:geo as map provider for WhereYouGo (such as for Locus at this moment) sounds nice for a soft start of integration. However it does not help with overcoming obstacles in WhereYouGo itself, like refactoring to new targetSDK, etc. So I doubt it will be meaningful.

So either integrate completely or leave it. 
@eddiemuc Can you give an idea, of how playing a cartridge could look like in c:geo according to your idea?

**Comment by eddiemuc** [comment]
@Lineflyer as the past has shown I am not really an expert for UX questions :-)

My simplified idea would be to take over WYG "as is" at first. 
* Todays WYG start screen is removed (preferences are integrated into settings, map viewer and gps viewer is already available in c:geo)
* WYG button is Integrated to c:geo main screen. If clicked then same happens as if clicking "start" button on today's WYG start screen (searches cartridges + asks to select one for play)
* If a cartridge is currently played but the WYG Dialogs are left to do other stuff in c:geo, cartridge play state can always be reached by using this icon again
* While a cartridge is being played, all relevant items are drawn into the live map (in addition to caches)
* Users might configure if certain cartridge events shall trigger an automatic change to WYG screen

From that point forward numerous additions are possible e.g.
* Redesign of WYG Dialogs
* Download/mgmt of cartridges from within app
* Other file structure for cartridges (place in public folder)
" "Shortcuts" as needed (e.g. maybe from map viewer to played cartridge)

Btw: we might even think about dropping the name "Whereyougo" inside c:geo and call it something like "c:geo Whereigo Player". Would be more consistent. But I would in any case place somewhere prominent where this came from.



This is totally up for discussion if

**Comment by moving-bits** [comment]
After a quick glance: Sounds good to me. Two additional ideas:
- We probably don't need an extra button on main screen, but could start / return to a running cartridge by pressing the WIG icon on the cache detail page.
- WIG state for a running cartridge could be persisted as JSON to the geocache table, similar to what we do for calculated waypoints.

**Comment by SchoolGuy** [comment]
Okay so my two cents as a WhereYouGo contributor:

I would vote for two apps:

- WhereYouGo is a Wherigo Player. Nothing more.
- c:geo is a Geocaching application.

If we do want to integrate WhereYouGo, we possibly want to integrate Mystery solving tools like GCC or other helpers as well. In the end we would have feature creep I guess.

Regarding stuff like the SAF and other things... Don't split it. Try to stick to the Google defaults as much as possible. This minimizes the migration efforts gives us lots of help because we are basically copy pasting from the Android Dev Docs then. Also I guess this means less code as we stick to what we need exactly.

The only exception I would make is the mapsforge support. This would mean a tiny small library. I wonder that no one has made a library but we have very extensive boilerplate code. See [here](https://github.com/mapsforge/mapsforge/tree/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android).

**Comment by Lineflyer** [comment]
I wanted to post an update here:
Seeing the fact, that more and more obstacles and efforts need to be taken to maintain WhereYouGo I meanwhile got to the opinion, that integrating it into c:geo might raise the chance to keep it alive as Android WheriGo player.

This is simply reasoned by the fact, that the current state of WhereYouGo is well behind everything:
- Outdated mapsforge
- Outdated GMaps
- Problems with online tile providers
- Not fit for targetSDK 30

Maybe it could really be a more realistic project to integrate it into c:geo and use the synergies provided. 
Would it be possible to do it in a plugin approach (meaning to provide a dedicated plugin install for interested users). That would IMHO avoid to make it just "part of c:geo" and would still allow a dedicated Playstore entry, separate crash reports and overall some kind of indepence.

**Comment by moving-bits** [comment]
FYI: We have a review of Google Play from today, asking for an integration of WhereYouGo in c:geo...

**Comment by eddiemuc** [comment]
Hehe😀

**Comment by Lineflyer** [comment]
With newer devices and older devices upgrading to more recent Android versions WhereYouGo becomes more and more restricted in its feature set (as no SAF means no offline maps and only a very restrictive way of accessing cartridge files).

As it seems unlikely that the big amount of work will be done on WhereYouGo (adding SAF, updating Mapsforge) it seems more realistic to again think about an integration into c:geo.

Anyone here interested in developing such approach and discussing the best method (see above my proposal about having it as plugin app)? Any other thoughts about that.
Its really a pity, that one of the few Android Whereigo apps (and the only open source one) is no longer fully usable.

**Comment by eddiemuc** [comment]
Update: since we introduced the navigation bar meanwhile and removed the old c:geo start screen, my suggestion would be now to replace the "nearby" action with a "Wherigo" action (and move the "nearby" search into the "search" activity)

**Comment by MagpieFourtyTwo** [comment]
Although I don't even use the Nearby button at all (IMO the map is the way better place to search for stuff, especially with the filters), thus would welcome a replacement, I don't really see the point in replacing it with a special cache type.

Or did you already integrate WhereIGos into c:geo and just want to tell us this way? ;)

**Comment by eddiemuc** [comment]
> Or did you already integrate WhereIGos into c:geo and just want to tell us this way? ;)

Unfortunately not :-(

**Comment by fm-sys** [comment]
@MagpieFourtyTwo you can already replace the nearby button via the appearance settings.

---
But I have to agree, why would we need a WIG launch button at all?

The original WhereYouGo Dashboard provides Start, Map, Location and Settings.
Settings should be merged with our settings, Map functionality should be merged with our map, Location can be removed or linked to the GPS info we show on our dashboard, and last but not least Start is IMHO not necessary anymore as we will allow launching WIGs via the cache details.

Only downside would be that a currently running WIG cannot be reached easily. But this could be solved by showing a system notification and link to the WIG if it gets clicked. 

BTW, seeing that the WhereYouGo app has an average rating of 1,5 stars and does not really receive regular updates, UI is totally outdated, and users claim that core functionality is not working, I think it's may be time to drop that app and include WIG functionality into cgeo.

Some developer need to be interested to work on it though...

**Comment by MagpieFourtyTwo** [comment]
> you can already replace the nearby button via the appearance settings.

Not only you're right, I even have already done so.  Completely forgot about it.  ;)

However - I (meanwhile) completely agree with integrating WYG into c:geo.  Wasn't really a friend of this idea, as it means to increase c:geo's complexity even more (and invent a completely new level of error-proneness), but since "others" have already done so, it now seems to be without alternative, cause c:geo shouldn't lag behind the "competition" ...

WRT the "error-proneness": WIGs are not really well known as rock solid. There are a lot of problems with crashing or simply bad programmed cartridges. And one must be very aware of that cartridge problems are not c:geo problems - so there may be a reasonably high risk that c:geo's reputation may suffer just because of bad cartridges ... a fact that therefore should always be made clear to users in the event of problems ...

**Comment by eddiemuc** [comment]
> > Some developer need to be interested to work on it though...
> 
> @SchoolGuy had started a POC some months ago, see [cgeo/WhereYouGo#73](https://github.com/cgeo/WhereYouGo/issues/73) Don't know the current state, though.

I believe this is an attempt to update mapsforge version used by WhereYouGo, not an integration of Wherigo into c:geo?

**Comment by eddiemuc** [comment]
> But I have to agree, why would we need a WIG launch button at all?

I believe IF c:geo provides a Whereigo implementation then it should become a "first-class-citizen". When I play whereigos in parallel to "normal" geocaching I have the need to switch to active running whereigo quite often. And I don't want to remember the location of the wherigo cache and find it again (on map or list) to go back to my active running game. 

Besides, if someone really doesn't want Wherigo in navigation bar, then the user can remove/replace it there.

> Some developer need to be interested to work on it though...

I am definitely interested, but only if functionality becomes a default "first-class-citizen" (because otherwise it would not be usable by myself :-( )

**Comment by fm-sys** [comment]
> When I play whereigos in parallel to "normal" geocaching I have the need to switch to active running whereigo quite often. And I don't want to remember the location of the wherigo cache and find it again (on map or list) to go back to my active running game.

I totally agree that it should be handled as a first class citizen. We should definitely allow WIG to be placed in the bottom navigation, however we IMHO shouldn't do it by default. There are so many geocachers who only play WIGs rarely or even not at all. 

That's why I suggested an additional android notification "You have started to play a WIG. Tap to resume." for the ones who don't want to have it in the bottom nav.

**Comment by moving-bits** [comment]
> > > Some developer need to be interested to work on it though...
> > 
> > @SchoolGuy had started a POC some months ago, see [cgeo/WhereYouGo#73](https://github.com/cgeo/WhereYouGo/issues/73) Don't know the current state, though.
> 
> I believe this is an attempt to update mapsforge version used by WhereYouGo, not an integration of Wherigo into c:geo?

It started as such, but discussion turned towards integration into c:geo [in the end](https://github.com/cgeo/WhereYouGo/issues/73#issuecomment-1646105216). But anyway, I haven't heard from him since then, don't know if he had found the time to dive deeper into it. So - if you have time and would be willing to try an integration of WhereYouGo into c:geo, this would be great news, IMHO, and I'm sure we will find a good way for a UI integration whits fits for both WIG-players and non WIG-players.

**Comment by geocachermgo** [comment]
When its integrated I would still like to have the possibility to switch between WIG and cgeo e.g. to add waypoints.

**Comment by fm-sys** [comment]
@capoaira AFAIK you were quite active in WIG development back then. Do you know of a cartridge using GIFs for testing purposes?

**Comment by capoaira** [comment]
> @capoaira AFAIK you were quite active in WIG development back then. Do you know of a cartridge using GIFs for testing purposes?

No, unfortunately not. I tried it a few times to build one with a GIF myself, but Urwigo don't support GIFs well, so I decide to don't use GIFs.

**Comment by eddiemuc** [comment]
I merged a first very drafty version into master. I added a short manual into our wiki on how to use it: https://github.com/cgeo/cgeo/wiki/Use-the-Draft-Wherigo-Player

**Comment by moving-bits** [comment]
> I merged a first very drafty version into master.

Thanks for all your work on integrating a Wherigo player into c:geo! I've played a bit with it today, trying to start Waldmeister's "Reverse Wherigo" cartridge. Some observations:
- opening the cartridge worked fine, as did the language selection (which is an text input element)
- selection of "unit" is missing the actual selection UI element (should be a selection list)
- same for the next question "Do you want to play a game or create a new one"; only "ok" button is given, but not the actual selection list
- the "game" ends here, you don't reach the screen where you can enter the start code)
- only possible follow-up action was to return to the Wherigo activity, which only showed some buttons (greyed-out "open", hightlighted "store" and "stop"), but no inventory, no actions etc.

Additionally, when compared this to playing the cartridge with WhereYouGo, the opening screen (with a short explanation for this cartridge and a small graphic) is missing.

Only then I realized that I had downloaded the "Pocket PC" version of the cartridge (as I'm used to since years), so I restarted with a "Garmin Colorado" version. But the results were the same. (Haven't checked the .gwc files' contents, but I assume the difference may be only the way images are compressed etc.?)

Link to Waldmeister's cartridge is: https://www.wherigo.com/cartridge/details.aspx?CGUID=dcdcd2ff-c171-4487-93bc-678f6d03ac4f

**Comment by eddiemuc** [comment]
> > I merged a first very drafty version into master.
> 
> Thanks for all your work on integrating a Wherigo player into c:geo! I've played a bit with it today, trying to start Waldmeister's "Reverse Wherigo" cartridge. Some observations:

Thanks a lot for this valuable feedback. I will get into all of it as time permits. Since I tested with one Wherigo only so far, it is kind of expected that a lot of stuff still comes up. Guess I will make the Waldmeister cartridge my new test object :-) 

> Only then I realized that I had downloaded the "Pocket PC" version of the cartridge (as I'm used to since years), so I restarted with a "Garmin Colorado" version. 

I tested with a Garmin Colorado, this is why I wrote that. MIght be as well that the Pocket PC version is better suited. I currently don't know which one is downloaded and used by WheYouGo - this version should become the main target. Is WhereYouGo using the Pocket PC version?

**Comment by eddiemuc** [comment]
> > Is WhereYouGo using the Pocket PC version?
> 
> `DownloadCartridgeTask` sets `ctl00$ContentPlaceHolder1$uxDeviceList` to `4`, which is the value for "Pocket PC Device", so yes, technically, but I don't know if there is any functional or encoding difference in the files (except maybe image compression or so). Both are ".gwc" files.

Then let's continue to use that version. I updated the wiki page accordingly

**Comment by eddiemuc** [comment]
> Link to Waldmeister's cartridge is: https://www.wherigo.com/cartridge/details.aspx?CGUID=dcdcd2ff-c171-4487-93bc-678f6d03ac4f

I am hitting some problem with this cartridge. It seems it performs some sort of "GPS spoof detection" (called "teste GPS-Empfang" in the cartridge). While I understand that in general, I don't know how to test that cartrdige in my emulator now... 
Anyone has an idea how to "pass" this "test" in the emulator?

**Comment by moving-bits** [comment]
You may want to [contact the author](https://www.geocaching.com/p/?code=PR3X6K0), he can probably give you some more info on the technical details.

**Comment by Lineflyer** [comment]
I have just enabled the feature and will for sure test it.
I do have some WheriGo in my home zone which are still to be played.  Will try to provide feedback as time permits.

**Comment by jakedot** [comment]
Audio is not yet supported, right?

**Comment by eddiemuc** [comment]
> Audio is not yet supported, right?

No

**Comment by eddiemuc** [comment]
> Audio is not yet supported, right?

Not yet supported.

To implement this, it would be great to have a wherigo cartridge using audio. Does anyone know of a good example cartridge for this?

**Comment by MagpieFourtyTwo** [comment]
[2016 hin und 2017 zurück](https://coord.info/GC6WPQX)

**Comment by MagpieFourtyTwo** [comment]
Even more [over here](https://www.google.com/search?q=WherIGo+Kopfhörer+site%3Ageocaching.com)  ;)

**Comment by geocachermgo** [comment]
I do know if its good but [Feuersturm](https://coord.info/GC1P1JW) had also sound if I remember right.

**Comment by eddiemuc** [comment]
@moving-bits or anyone else: do you have any pointers for me on how to integrate Wherigo cartridge download within our download services in c:geo?

Especially: how can I get notified when download is completed?

**Comment by moving-bits** [comment]
Hi @eddiemuc,

here are some pointers, feel free to ask if more is required:

- all downloaders are located in /downloader subfolder
- there are a couple of abstract classes for different types of download items; most of them are designed to do some parsing of a source page to provide a list of available downloads
- have a look at `Routing.java`, search for `DownloadConfirmationActivity`, for directly triggering a download without further parsing of some source page
- trigger a download by calling `DownloaderUtils.triggerDownload(...) `, providing a unique `Download.DownloadType` for Wherigo cartridges
- c:geo gets notified for a successful (or failing) download in `DownloadNotificationReceiver`, which creates a worker job for copying the received file into its destination folder. See (and probably adapt, if required) `ReceiveDownload` class
- `onSuccessfulReceive` (inherited from `AbstractDownloader`) may be a convenient place to trigger follow-up actions

**Comment by eddiemuc** [comment]
Thanks @moving-bits for quick and comprehensive reply!

Looking into WhereYouGo code, download of cartridge has to be initiated with a POST request. After googling a bit, it seems that Android's DownloadManager is not able to handle this. Looks like I need to keep the download "hand-crafted"... :-(

**Comment by jakedot** [comment]
When testing with a cartridge I have in development currently, I'm getting error Toasts when I move into a zone. Is that implemented yet? Or maybe related to FakeGPS?

Attached please find my cartridge.

[ica.zip](https://github.com/cgeo/cgeo/files/14843568/ica.zip)

**Comment by jakedot** [comment]
Has there been any development on wherigo integration recently?

**Comment by eddiemuc** [comment]
> Has there been any development on wherigo integration recently?

no

**Comment by eddiemuc** [comment]
With the merging of PR https://github.com/cgeo/cgeo/pull/16161 we now have a function-complete experimental wherigo player integrated into c:geo

Start the player from a Wherigo Detail page by clicking on the now-added second "Wherigo-Icon":
![text](https://github.com/user-attachments/assets/03918999-f5a3-4676-9895-04239df10ddd)

Further instructions can be found here:
https://github.com/cgeo/cgeo/wiki/Use-the-Draft-Wherigo-Player

There suerly will be tons of remarks and bugs, and GUI has to be discussed amongst the team. However, since initial version is now available I will close this rather large issue for now and suggest to open dedicated issues for topics that will come up when the player is firstly used.

**Comment by MagpieFourtyTwo** [comment]
We had our first use today - admittedly, with a rather easy cartridge, but it worked pretty well.

Only "real" issue we had (apart from understanding the kind of misleading icons 🙃) was coming back to the WIG once you've returned to c:geo's home screen or even left c:geo by pressing ``Back`` too often. 🙄😜
In this case, although the mentioned notification was still there (and the player obviously still running), it's not possible to get back to the WIG by clicking the notification. Sometimes after the notification click just nothing happened, sometimes c:geo's home screen was shown, but we did not manage to get back into the WIG to correctly terminate the player. Most regretfully I can now reproduce just one scenario: Play the WIG, then press ``Back`` till you're in c:geo's home, then click the notification ... and you will just be put back to c:geo's home page. To get back to the player, you have to press ``Back`` once more, to get to Android's home screen, and then the notification tap will take you back to the WIG (which for sure did not work in the field neither, but now is does ...)

Only way then was, to open the WIG listing and press the "new" WIG icon. After pressing ``New game`` the message appears that there is a WIG currently running, asking if you want to end it - then press ``Cancel`` and you're back. Guess this should be a bit more intuitive ... 😎

Ah, and the button ``Locate on center (10S)`` - what is this really for?  Just to see the message of the waypoint area?  Guess this is for debug purposes ...?  Still don't get what ``10S`` may stand for ... 

As for the icons: For "Chose a cartridge", I would show a play button or so.  The current one looks more like "upload".
And the "folder", which stands for "Download" may perhaps be better s/th with a downwards arrow, wouldn't it?
Ah, yeah, the "autosave" - can resp must it in any way be chosen manually or does the player do it just automatically?

Ah, just got the second scenario reproduced, when it was not possible to come back to the WIG once you're in Android's home screen: This happens, if you leave the WIG by pressing ``Back`` till you are in c:geo's home screen, and then you press Android's Home button - if you now press the notification, you will just get back to c:geo's home. ;)

Nevertheless, very good work, thx a lot! This will give WIGs a new life!  :)

**Comment by MagpieFourtyTwo** [comment]
Ok, to be honest: We just chose the WIG we played as our today's daily, cause we were curious how the player will behave.
And we have not been disappointed!  :)

**Comment by eddiemuc** [comment]
One more thing: we had a discussion about how the WherigoPlayer may be accessed. I integrated both quick launch and custom menu item to get to the player but I assume most users won't configure this:

![Screenshot_2024-10-05-08-30-46-78_a916ceb0c8a82619faeb76fe6b8ae997.jpg](https://github.com/user-attachments/assets/9cd40fd1-6379-46f7-898e-555a18decf43)

I will fix the bug found by @MagpieFourtyTwo wrt click on notification of course. Leaving this aside, can I pursue the team to put a more permanent link to the player into c:geo's GUI? Maybe a preconfigured or even permanent quick launch button?

**Comment by moving-bits** [comment]
> but I am seriously thinking about offering it to users as well to retrigger "lost" messages in the field. I had often the situation that I had to return to a previous zone just because some previous information was lost. But this is discussable, what does the team think?

Sounds useful to me (but could be considered cheating for some cartridges 😀). If you decide to keep it, it could be named "Revisit zone (10s)".

**Comment by MagpieFourtyTwo** [comment]
I like the idea of having an additional bottom menu entry for the player, but in smaller displays space may be an issue.

Thus I would at least set a default for the quick launch icon - with which BTW you can even provide a quick fix for the notification click problem, cause when the notification just brings you to c:geo's home screen, another click on the WIG icon brings you right back into the WIG.  ;)

**Comment by fm-sys** [comment]
A quick launch button for something many users won't use isn't a good idea IMHO. Especially, as it requires two clicks instead of one for those who want to use the feature. 

> with which BTW you can even provide a quick fix for the notification click problem, 

Well, let's rather fix the root cause instead of implementing workarounds :) 

Btw: Signal messenger shows a box below the action bar if a voice message is currently running. This idea could be used for WIGs as well...

![signal-2024-10-05-09-34-44-276](https://github.com/user-attachments/assets/80af038e-aa9c-4873-95e3-3c688bb76563)

**Comment by MagpieFourtyTwo** [comment]
> "Revisit zone (10s)".

Good key word, which led me to an idea WRT cheating (which was in fact the first thing I had in mind, too, after testing it): What if this button is only visible *after* you have visited this station in the WIG?

**Comment by eddiemuc** [comment]
> > "Revisit zone (10s)".
> 
> Good key word, which led me to an idea WRT cheating (which was in fact the first thing I had in mind, too, after testing it): What if this button is only visible *after* you have visited this station in the WIG?

Had the same thought. This would have an interesting impact: we would need our own "save file" mechanism on top of the one from the OpenWIG engine. Because we would need to remember which Zones the user has already accessed. And also this might be considered "cheating": many wherigos set previous Zones to "invisible" at some point. If we let users "see" them nevertheless, well, might not what the Wherigo author intended. 

I feel we have to have a discussion about what is considered "cheating" :-)

**Comment by eddiemuc** [comment]
> A quick launch button for something many users won't use isn't a good idea IMHO. Especially, as it requires two clicks instead of one for those who want to use the feature. 

That's the key of discussion: do we consider Wherigos a c:geo side feature only used by very few users or a main feature?

The two other Geocaching apps I know offering Wherigo support place this very prominent (one even in the title).

Can I at least have an overflow menu entry on home screen?

**Comment by eddiemuc** [comment]
> Well, let's rather fix the root cause instead of implementing workarounds :) 

Do not despair, this will be fixed regardless of the "Wherigo Player access" discussions outcome :-)

**Comment by eddiemuc** [comment]
> Btw: Signal messenger shows a box below the action bar if a voice message is currently running. This idea could be used for WIGs as well...
> 
> ![signal-2024-10-05-09-34-44-276](https://github.com/user-attachments/assets/80af038e-aa9c-4873-95e3-3c688bb76563)

Would be not as easy to implement as you might think, because if a wherigo plays an audio, user might be in any screen in c:geo. Or do you mean "just show this in wherigo activity"?

**Comment by MagpieFourtyTwo** [comment]
WIGs have been a topic on every single event we visited lately.  Only me knows about a dozen users who stumble to get WhereYouGo running with Android >12. And they keep asking how to do it ...

In yesterday's logs I mentioned the alpha test of "a new WIG player" in a side note (TBH as a test balloon ;) ) and immediately got two email reactions about how curious they are about a new player ...

Thus I would consider the WIG theme a way bigger one than one may expect on first sight. And would highly vote for at least the quick launch button, which additionally allows the access of downloaded WIGs without first having to search for the listing.

**Comment by fm-sys** [comment]
> Would be not as easy to implement as you might think, because if a wherigo plays an audio, user might be in any screen in c:geo. Or do you mean "just show this in wherigo activity"?

Ah, well, I should have explained it further 😅

My idea was not related to audio output of WIGs.

At any time a WIG is currently running, we could show such box in all (top-level?) activities with the text "_a WIG is currently running... Click to resume_" or similar. 
Users who need to access a currently running WIG can do this very easily, while those who don't aren't disturbed.

**Comment by fm-sys** [comment]
Other relevant features such as PQs, bookmark lists, etc are listed in the 3-dot-menu of home. IMHO WIGs are at a similar level of importance, so I would suggest to add it there. Users can add a WIG quick launch button manually if they want and need, as with any other feature mentioned above.

**Comment by eddiemuc** [comment]
> Other relevant features such as PQs, bookmark lists, etc are listed in the 3-dot-menu of home. IMHO WIGs are at a similar level of importance, so I would suggest to add it there. Users can add a WIG quick launch button manually if they want and need, as with any other feature mentioned above. 

Done and merged: https://github.com/cgeo/cgeo/pull/16167

**Comment by eddiemuc** [comment]
> My idea was not related to audio output of WIGs.
> 
> At any time a WIG is currently running, we could show such box in all (top-level?) activities with the text "_a WIG is currently running... Click to resume_" or similar. 
> Users who need to access a currently running WIG can then do this very easily, while those who don't aren't disturbed.

Phew, only if this can be deactivated with a setting. Because I (and others) for sure will have wherigo as a custom menu entry in bottom navigation and such a feature would just be a waste of space.

Please create an issue for this if you want us to pursue it

**Comment by rouroni1** [comment]
@eddiemuc 
> Autosaves are triggered automatically by the game engine OpenWIG. Can't tell you the conditions for sure though, this is not documented :-) Should be same behavior as in WhereYouGo.

Autosaves are triggered in the programming of the cartridge.
I use to add an autosave when the player enters a zone, or when grabs a object in the game.
The main reason that I do that, is beacause whereyougo sometimes crashed out of the blue, and if there weren't  any manual saves the player only had the option to start everything again.

![image](https://github.com/user-attachments/assets/8aa41507-0475-4cd8-bd6c-0520e75fb04a)

**Comment by rouroni1** [comment]
@eddiemuc @MagpieFourtyTwo @moving-bits 
> > but I am seriously thinking about offering it to users as well to retrigger "lost" messages in the field. I had often the situation that I had to return to a previous zone just because some previous information was lost. But this is discussable, what does the team think?
> 
> Sounds useful to me (but could be considered cheating for some cartridges 😀). If you decide to keep it, it could be named "Revisit zone (10s)".

As a creator of wherigos in urwigo, this option is great for testing, but... is excelent for cheating.

Example, in the wherigo that I'm testing, it's the game of logic Wolf, Sheep and lettuce, the player must go to the same location at least 4 times, so the ability to revisit the zone without realy going there for me it's a no no...

**Comment by eddiemuc** [comment]
> > > but I am seriously thinking about offering it to users as well to retrigger "lost" messages in the field. 
> > 
> > Sounds useful to me (but could be considered cheating for some cartridges 😀). 
> As a creator of wherigos in urwigo, this option is great for testing, but... is excelent for cheating.

I restricted it to debug mode for now.

And how you enter debug mode for a cartridge is not documented currently :-)

**Comment by rouroni1** [comment]
> > @capoaira AFAIK you were quite active in WIG development back then. Do you know of a cartridge using GIFs for testing purposes?
> 
> No, unfortunately not. I tried it a few times to build one with a GIF myself, but Urwigo don't support GIFs well, so I decide to don't use GIFs.

@eddiemuc 

I made a quick test and Gift files work with c:geo Wherigo Player 🙂


https://github.com/user-attachments/assets/1bc01be9-626b-4a92-b736-ad27b07272a7

[Teste_Gif_2.zip](https://github.com/user-attachments/files/18140586/Teste_Gif_2.zip)

**Comment by moving-bits** [comment]
@eddiemuc 
We got feedback on support today w/r to the integrated Wherigo player, I thought I would share that with you:

Hi,

habt ihr super gemacht!

 GC3P85G

Dieser Whereyougo diente heute zum Test der neuen Funktion bei [c.geo](http://c.geo/).
Ich bin total begeistert alles hat super funktioniert und die seperate App ist nun überflüssig.

(see ticket 469136)
